Firestore/core/src/firebase/firestore/util/async_queue.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_ASYNC_QUEUE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_ASYNC_QUEUE_H_
+
+#include <atomic>
+#include <chrono>  // NOLINT(build/c++11)
+#include <functional>
+#include <memory>
+
+#include "Firestore/core/src/firebase/firestore/util/executor.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+/**
+ * Well-known "timer" ids used when scheduling delayed operations on the
+ * AsyncQueue. These ids can then be used from tests to check for the
+ * presence of delayed operations or to run them early.
+ */
+enum class TimerId {
+  /** All can be used with `RunDelayedOperationsUntil` to run all timers. */
+  All,
+
+  /**
+   * The following 4 timers are used in `Stream` for the listen and write
+   * streams. The "Idle" timer is used to close the stream due to inactivity.
+   * The "ConnectionBackoff" timer is used to restart a stream once the
+   * appropriate backoff delay has elapsed.
+   */
+  ListenStreamIdle,
+  ListenStreamConnectionBackoff,
+  WriteStreamIdle,
+  WriteStreamConnectionBackoff,
+
+  /**
+   * A timer used in `OnlineStateTracker` to transition from
+   * `OnlineStateUnknown` to `Offline` after a set timeout, rather than waiting
+   * indefinitely for success or failure.
+   */
+  OnlineStateTimeout,
+};
+
+// A serial queue that executes given operations asynchronously, one at a time.
+// Operations may be scheduled to be executed as soon as possible or in the
+// future. Operations scheduled for the same time are FIFO-ordered.
+//
+// `AsyncQueue` wraps a platform-specific executor, adding checks that enforce
+// sequential ordering of operations: an enqueued operation, while being run,
+// normally cannot enqueue other operations for immediate execution, called
+// "nesting" (but see `EnqueueAllowingNesting`).
+//
+// `AsyncQueue` methods have particular expectations about whether they must be
+// invoked on the queue or not; check "preconditions" section in comments on
+// each method.
+//
+// A significant portion of `AsyncQueue` interface only exists for test purposes
+// and must *not* be used in regular code.
+class AsyncQueue {
+ public:
+  using Operation = internal::Executor::Operation;
+  // A more-or-less arbitrary unit of time for scheduling operations in the
+  // future.
```

Comment:
Removed. I try to avoid unnecessary comments (in this case, I was trying to explain why it's milliseconds as opposed to e.g. seconds or microseconds, but if someone else reading this thinks it doesn't add value, I'm all for removing it).